### PR TITLE
support training pipeline

### PR DESCRIPTION
**Pull request by JacoCheung**
## Description

1. Enable datadist and prefetch pipeline. Users can specify the pipeline type as `none`, `native` or `prefetch` to use the pipeline. 
2. Unify train/eval api with pipeline (strip off the metric module)
3. refactor training utils
4. Improve the nvtx_op utils
5. add e2e pipeline test.
## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rapids-repo-template/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: examples/hstu/prefetch_pipeline_demo.gin [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+BenchmarkDatasetArgs.feature_args = [
+    @item_and_action_feature/FeatureArgs(),
+]
+
+item_embedding/EmbeddingArgs.feature_names = ['item']
+item_embedding/EmbeddingArgs.table_name = 'item'
+item_embedding/EmbeddingArgs.item_vocab_size_or_capacity = 5000000 # gross 50M embedding rows
+item_embedding/EmbeddingArgs.sharding_type = 'model_parallel'
+item_embedding/EmbeddingArgs.allowed_compute_kernels = ['fused_uvm_caching',"dense"]
```
Comment (shijieliu) [inline examples/hstu/prefetch_pipeline_demo.gin]: what will happen if user use `allowed_compute_kernels = ["dense"]` and `pipeline_type =prefetch`?

If there is a coupling between compute_kernels and pipeline_type, how about not allowing users to configure compute_kernels?

File: examples/hstu/pretrain_gr_ranking.py [hunk 2]
```diff
@@ -129,14 +139,32 @@ def main():
...
+        )
+        train_with_pipeline(
+            pipeline,
+            trainer_args,
+            train_dataloader,
+            test_dataloader,
+            dense_optimizer,
+        )
+    else:
+        train(
```
Comment (shijieliu) [inline examples/hstu/pretrain_gr_ranking.py]: can we move the train under folder pipeline and merge ranking and retrieval train?

File: examples/hstu/modules/embedding.py [hunk 3]
```diff
@@ -298,38 +331,49 @@ def create_embedding_collection(configs):
...
-        if self._data_parallel_embedding_collection is not None:
-            with torch.cuda.stream(self._side_stream):
-                dp_embeddings = self._data_parallel_embedding_collection(kjt)
-            torch.cuda.current_stream().wait_stream(self._side_stream)
-            embeddings = {**mp_embeddings_awaitables.wait(), **dp_embeddings}
-        else:
-            embeddings = mp_embeddings_awaitables.wait()
-        return embeddings
+        # Use wrapped helper functions for better traceability
+        mp_embeddings = _get_mp_embeddings(
```
Comment (shijieliu) [inline examples/hstu/modules/embedding.py]: do we need those code if we are rewriting model from nn.module? Otherwise we can revert the changes in this file

File: examples/hstu/distributed/sharding.py [hunk 4]
```diff
@@ -380,6 +394,8 @@ def make_optimizer_and_shard(
     sparse_optimizer_param: OptimizerParam,
     dense_optimizer_param: OptimizerParam,
     dynamicemb_options_dict: Dict[str, DynamicEmbTableOptions] = {},
+    allowed_compute_kernels: Dict[str, List[str]] = {},
+    enable_prefetch_pipeline: bool = False,
```
Comment (shijieliu) [inline examples/hstu/distributed/sharding.py]: can we decide `allowed_compute_kernels ` by enable_prefetch_pipeline?

File: examples/hstu/utils.py [hunk 5]
```diff
@@ -521,6 +542,63 @@ def train(
...
+            cur_td = gpu_timer.elapsed_time() - last_td
+            print_rank_0(
+                f"[train] [iter {train_iter}, tokens {int(tokens_logged.item())}, elapsed_time {cur_td:.2f} ms]: loss {reporting_loss[0] / reporting_loss[1]:.6f}"
+            )
+            last_td = cur_td + last_td
+            tokens_logged.zero_()
+
+        if train_iter > 0 and train_iter % trainer_args.eval_interval == 0:
+            pipeline._model.eval()
+            evaluate(
```
Comment (shijieliu) [inline examples/hstu/utils.py]: do we need to change evaluate as well if we enable pipeline? since model is rewritten.

File: examples/hstu/modules/embedding.py (hunk 3 above)
Comment (JacoCheung) [reply examples/hstu/modules/embedding.py]: You're right. We now rewrite model bypassing fx.Tracer, so we don't need those changes.

File: examples/hstu/training/utils.py [hunk 6]
```diff
@@ -372,153 +213,39 @@ def create_dynamic_optitons_dict(
...
+    "native": ["dense"],
+    "none": [],
+}
+_sharding_type_to_allowed_compute_kernels = {
+    "data_parallel": _pipeline_type_to_data_parallel_allowed_compute_kernels,
+    "model_parallel": _pipeline_type_to_model_parallel_allowed_compute_kernels,
+}
+
+
+def create_embedding_compute_kernels(
```
Comment (shijieliu) [inline examples/hstu/training/utils.py]: can we make this function as 

```
def create_embedding_compute_kernels(
    model: nn.Module,
    pipeline_type: str,
) -> Dict[str, List[str]]:
```

and use it inside apply_dmp?

Comment (JacoCheung) [reply examples/hstu/training/utils.py]: I removed the functions and get the compute type inside planner. Please review! @shijieliu

File: examples/hstu/training/gin_config_args.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    profile_step_end: int = 200
+    # ==nsys args==
+    max_train_iters: Optional[int] = None
+    max_eval_iters: Optional[int] = None
+
+    # ckpt args
+    ckpt_save_interval: int = -1  # -1 means not save ckpt
+    ckpt_save_dir: str = "./checkpoints"
+    ckpt_load_dir: str = ""
+    pipeline_type: str = "none"  # none, native, prefetch
```
Comment (shijieliu) [inline examples/hstu/training/gin_config_args.py]: how about make this default value as "native"? did you test the accuracy for kuairand and movielens dataset?

Comment (JacoCheung) [reply examples/hstu/training/gin_config_args.py]: > Set this default value as "native"

Done.

> did you test the accuracy for kuairand and movielens dataset 

The results looked fine. Can you help check if it makes sense? @shijieliu 
[native_pipeline_kuairand_1K.log](https://github.com/user-attachments/files/20941732/native_pipeline_kuairand_1K.log)

[native_pipeline_kuairand_27K.log](https://github.com/user-attachments/files/20941840/native_pipeline_kuairand_27K.log)
